packages/api/internal/handlers/sandbox_metrics.go

```diff
@@ -0,0 +1,127 @@
+package handlers
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/launchdarkly/go-sdk-common/v3/ldcontext"
+	"go.uber.org/zap"
+
+	"github.com/e2b-dev/infra/packages/api/internal/api"
+	"github.com/e2b-dev/infra/packages/api/internal/auth"
+	authcache "github.com/e2b-dev/infra/packages/api/internal/cache/auth"
+	clickhouse "github.com/e2b-dev/infra/packages/clickhouse/pkg"
+	featureflags "github.com/e2b-dev/infra/packages/shared/pkg/feature-flags"
+	"github.com/e2b-dev/infra/packages/shared/pkg/logger"
+)
+
+func (a *APIStore) GetSandboxesSandboxIDMetrics(c *gin.Context, sandboxID string, params api.GetSandboxesSandboxIDMetricsParams) {
+	ctx := c.Request.Context()
+	ctx, span := a.Tracer.Start(ctx, "sandbox-metrics")
+	defer span.End()
+
+	team := c.Value(auth.TeamContextKey).(authcache.AuthTeamInfo).Team
+
+	flagCtx := ldcontext.NewBuilder(featureflags.MetricsReadFlagName).Build()
+	metricsReadFlag, err := a.featureFlags.Ld.BoolVariation(featureflags.MetricsReadFlagName, flagCtx, featureflags.MetricsReadDefault)
+	if err != nil {
+		zap.L().Error("error getting metrics read feature flag, soft failing", zap.Error(err))
+	}
+
+	if !metricsReadFlag {
+		zap.L().Debug("sandbox metrics read feature flag is disabled")
+		// If we are not reading from ClickHouse, we can return an empty map
+		// This is here just to have the possibility to turn off ClickHouse metrics reading
+
+		c.JSON(http.StatusOK, []api.SandboxMetric{})
+		return
+	}
+
+	start, end, err := getStartEndTime(ctx, a.clickhouseStore, team.ID.String(), sandboxID, params)
+	if err != nil {
+		a.sendAPIStoreError(c, http.StatusInternalServerError, fmt.Sprintf("error when getting metrics time range: %s", err))
+		return
+	}
+
+	step := calculateStep(start, end)
+
+	metrics, err := a.clickhouseStore.QuerySandboxMetrics(ctx, sandboxID, team.ID.String(), start, end, step)
+	if err != nil {
+		zap.L().Error("Error fetching sandbox metrics from ClickHouse",
+			logger.WithSandboxID(sandboxID),
+			logger.WithTeamID(team.ID.String()),
+			zap.Error(err),
+		)
+
+		a.sendAPIStoreError(c, http.StatusInternalServerError, fmt.Sprintf("error querying sandbox metrics: %s", err))
+		return
+	}
+
+	apiMetrics := make([]api.SandboxMetric, len(metrics))
+	for i, m := range metrics {
+		apiMetrics[i] = api.SandboxMetric{
+			Timestamp:  m.Timestamp,
+			CpuUsedPct: float32(m.CPUUsedPercent),
+			CpuCount:   int32(m.CPUCount),
+			MemTotal:   int64(m.MemTotal),
+			MemUsed:    int64(m.MemUsed),
+		}
+	}
+
+	c.JSON(http.StatusOK, apiMetrics)
+}
+
+// calculateStep determines the step size for metrics based on the time range.
+// The result should always contain less than 1000 points.
+func calculateStep(start, end time.Time) time.Duration {
```

Comment:
Not sure here if it should be just based on start and end - for paused and resumed sandbox it can result in only few data points, but on the other side, you probably won't be able to use more point in the plot it in the dashboard